pkg/config/manifest.go

```diff
@@ -139,72 +186,84 @@ func (od *OutputDefinition) Validate() error {
 	return nil
 }
 
-func (d *Dependency) Validate() error {
-	if d.Tag == "" {
-		return errors.New("dependency tag is required")
-	}
+type BundleOutput struct {
+	Name                string `yaml:"name"`
+	Path                string `yaml:"path"`
+	EnvironmentVariable string `yaml:"env"`
+}
 
-	if strings.Contains(d.Tag, ":") && len(d.Versions) > 0 {
-		return errors.New("dependency tag can only specify REGISTRY/NAME when version ranges are specified")
-	}
+type Steps []*Step
 
+func (s Steps) Validate(m *Manifest) error {
+	for _, step := range s {
+		if step != nil {
+			err := step.Validate(m)
+			if err != nil {
+				return err
+			}
+		}
+	}
 	return nil
 }
 
-func resolveParameter(pd ParameterDefinition) (string, error) {
-	pe := pd.Name
-	if pd.Destination == nil {
-		// Porter by default sets CNAB params to name.ToUpper()
-		return os.Getenv(strings.ToUpper(pe)), nil
-	} else if pd.Destination.EnvironmentVariable != "" {
-		return os.Getenv(pd.Destination.EnvironmentVariable), nil
-	} else if pd.Destination == nil && pd.Destination.Path != "" {
-		return pd.Destination.Path, nil
-	}
-	return "", fmt.Errorf("parameter: %s is malformed", pd.Name)
-
+type Step struct {
+	Data map[string]interface{} `yaml:",inline"`
 }
 
-func resolveCredential(cd CredentialDefinition) (string, error) {
-	if cd.EnvironmentVariable != "" {
-		return os.Getenv(cd.EnvironmentVariable), nil
-	} else if cd.Path != "" {
-		return cd.Path, nil
-	} else {
-		return "", fmt.Errorf("credential: %s is malformed", cd.Name)
+func (s *Step) Validate(m *Manifest) error {
+	if len(s.Data) == 0 {
+		return errors.New("no mixin specified")
+	}
+	if len(s.Data) > 1 {
+		return errors.New("more than one mixin specified")
 	}
-}
 
-func (d *Dependency) resolve() (map[string]interface{}, []string, error) {
-	sensitiveStuff := []string{}
-	depVals := make(map[string]interface{})
+	mixinDeclared := false
+	mixinType := s.GetMixinName()
+	for _, mixin := range m.Mixins {
+		if mixin == mixinType {
+			mixinDeclared = true
+			break
+		}
+	}
+	if !mixinDeclared {
+		return errors.Errorf("mixin (%s) was not declared", mixinType)
+	}
 
-	params := make(map[string]interface{})
-	depVals["parameters"] = params
-	// TODO: Populate dependency parameters lookup
+	if _, err := s.GetDescription(); err != nil {
+		return err
+	}
 
-	creds := make(map[string]interface{})
-	depVals["credentials"] = creds
-	// TODO: Resolve dependency credentials lookup, or remove it from the template language if it shouldn't be accessible
+	return nil
+}
 
-	outputs := make(map[string]interface{})
-	depVals["outputs"] = outputs
-	// TODO: Populate dependency output lookups
-	// TODO: Add outputs onto sensitive stuff
+// GetDescription returns a description of the step.
+// Every step must have this property.
+func (s *Step) GetDescription() (string, error) {
+	if s.Data == nil {
+		return "", errors.New("empty step data")
```

Comment:
I'm assuming this wasn't changed here, but I had the thought:  I wonder if we want to append `"Must minimally provide a step description"` or some such here to this error message?

Reply:
No code was changed, just moved into a separate file.